### PR TITLE
Rccl replayer

### PR DESCRIPTION
**Pull request by BertanDogancay**
* Update enqueue.cc to add additional info to collective logs.
* Develop rcclReplayer
* create a README file for the replayer

### CODE REVIEW

**Review by wenkaidu — APPROVED** [review]

File: tools/rccl_replayer/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+Depending on the MPI library used and your installation path, you may need to edit the Makefile and set the MPI_DIR path accordingly.
+
+
+## Usage
+
+After successfully building the replayer, you can run it using the following command:
+
+```bash
+    NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=COLL mpirun -np <numProcesses> ./rcclReplayer </path/to/logfile> <numGpusPerMpiRank>
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/README.md:58]: There's no need to use NCCL_DEBUG=INFO or NCCL_DEBUG_SUBSYS to run the replayer.

File: tools/rccl_replayer/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+Replace 'numProcesses' with the number of MPI processes you want to run during the replay, </path/to/logfile> with the path to the collective log file generated during your RCCL runs, and 'numGpusPerMpiRank' with the number of GPUs per MPI rank used in your application.
+
+Depending on the MPI library you use, you may need to modify the mpirun command accordingly. The flag NCCL_DEBUG_SUBSYS=COLL ensures that only collective log information is printed to the terminal.
+
+### Multi-Node Environment:
+
+If multiple nodes were used for your application, you can also replay the collective logs with using multiple nodes. See the following command:
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/README.md:67]: extra "with"

File: tools/rccl_replayer/rcclReplayer.cpp [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+                auto& rankVector = arrivalCounter[key];
+                if (rankVector.size() < numGlobalRanks) {
+                    rankVector.resize(numGlobalRanks);
+                }
+
+                // rankVector<int> in arrivalCount represents the rank information
+                // Count the number of tasks that are going to be executed by each rank. This is to validate the group call later on.
+                // Nom-Send/Recv rank counts (rankVector<int> elements) should be equal at the end, and for Send/Recv, all the elements of rankVector<int> should be equal to 0
+                if (strcmp(ncclFuncNames[ti.funcType], "Recv") == 0) {
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/rcclReplayer.cpp:133]: just use ti.funcType

File: tools/rccl_replayer/rcclReplayer.cpp [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+            
+            bool isp2p = (strcmp(std::get<0>(e.first), "Send/Recv") == 0);
+            if (!isp2p) maxVal = *std::max_element(e.second.begin(), e.second.end());
+            
+            // Validate all the ranks have required amount of collective call (task)
+            for (int i = 0; i < e.second.size(); i++) {
+                if (e.second[i] != (isp2p ? 0 : maxVal)) {
+                    std::string warning = (isp2p ? (e.second[i] > 0 ? "[WARN] Missing Recv" : "[WARN] Missing Send") : "[WARN] Missing " + std::string(std::get<0>(e.first)) ) 
+                            + " count=" + std::to_string(std::get<1>(e.first)) + " datatype=" + std::to_string(std::get<2>(e.first)) + " op=" + std::to_string(std::get<3>(e.first)) + " at rank [" + std::to_string(i) + "]";
+                    if(mpiRank == 0) printf("%s\n", warning.c_str());
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/rcclReplayer.cpp:154]: This is hard to follow with all the get<>.
Given e, just create named references to the various tuple fields, and use that in your code to improve readability
e.g.:
char* funcName = std::get<0>(e.first);
size_t count          = std::get<1>(e.first);
...

File: tools/rccl_replayer/rcclReplayer.cpp [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+
+                    gc.isValid = false;
+                }
+            }
+        }
+    }
+}
+
+std::pair<size_t, size_t> GetSize(TaskInfo taskInfo, int numGlobalRanks) {
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/rcclReplayer.cpp:163]: Add a comment about what this pair represents (#bytesSent, #bytesRecv)

File: tools/rccl_replayer/rcclReplayer.cpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+            break;
+        default:
+            printf("Error: unsupported collective\n");
+            exit(1);
+    }
+}
+
+void ReplayRccl(GroupCall& groupCall, std::vector<ncclComm_t> comms, std::vector<hipStream_t> streams,
+                                                                     int const localGpuOffset, int const numGpusPerMpiRank, int const firstGlobalRank, int const numGlobalRanks) {
+    std::vector<TaskInfo>& tasks = groupCall.rankData[localGpuOffset].tasks;
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/rcclReplayer.cpp:229]: This vector is indexed by localGpuOffset, which is wrong because rankData is indexed by globalRank.
However, this vector isn't even used.

File: tools/rccl_replayer/rcclReplayer.cpp [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+
+    void* sendbuff[numGpusPerMpiRank];
+    void* recvbuff[numGpusPerMpiRank];
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/rcclReplayer.cpp:232]: This looks wrong still.  Each task needs its own send/recv buffer for each GPU.
This should be a vector<vector<void*>> sendbuff(numGpusPerMpiRank), with the inner vector fluxuating with the number of tasks for that rank in the group call.

File: tools/rccl_replayer/rcclReplayer.cpp [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+    NCCLCHECK(ncclGroupStart());
+    for (auto rd : groupCall.rankData) {
+        for (int i = 0; i < numGpusPerMpiRank; i++) {
+	    int globalRank = firstGlobalRank + i;
+            if (globalRank == rd.first) {
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/rcclReplayer.cpp:238]: This is a very awkward way of iterating because there's an extra for loop/if statement.

For (int localIdx = 0; localIdx < numGpusPerMpiRank; localIdx++)
{
  int globalRank = firstGlobalRank + localIdx;
  RankData& rankData = groupCall.rankData[globalRank];
  ...

File: tools/rccl_replayer/rcclReplayer.cpp [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+
+                    if (task.inPlace) {
+                        numBytes.first = std::max(numBytes.first, numBytes.second);
+                        numBytes.second = numBytes.first;
+                    }
+                    
+                    // Set the device and allocate send/recv buffers
+                    HIPCALL(hipSetDevice(localGpuOffset + i));
+                    HIPCALL(hipMalloc(&(sendbuff[i]), numBytes.first));
+                    HIPCALL(hipMalloc(&(recvbuff[i]), numBytes.second));
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/rcclReplayer.cpp:251]: Memory leak because send/recv buff not accounting for per-task.
See above.

File: tools/rccl_replayer/rcclReplayer.cpp [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+    
+    // Synchronize devices
+    for (int i = 0; i < numGpusPerMpiRank; i++) {
+        HIPCALL(hipStreamSynchronize(streams[i]));
+    }
+
+    // Free device memory
+    for (int i = 0; i < numGpusPerMpiRank; i++) {
+        HIPCALL(hipFree(sendbuff[i]));
+        HIPCALL(hipFree(recvbuff[i]));
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/rcclReplayer.cpp:272]: Memory leak.  See above

**Review by gilbertlee-amd — CHANGES_REQUESTED** [review]

File: tools/rccl_replayer/Makefile [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+ifdef MPI_DIR
+main: rcclReplayer.cpp
+	/opt/rocm/bin/hipcc rcclReplayer.cpp \
+		-g \
+		-o rcclReplayer \
+		-I$(MPI_DIR)/ \
+		-I$(RCCL_DIR) \
+		-I$(RCCL_DIR)/include/rccl \
+		-I/opt/rocm/include/hip \
+				-L$(MPI_DIR)/lib \
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/Makefile:12]: spacing

File: tools/rccl_replayer/README.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+## Introduction
+
+Replayer is a dubugging tool designed to analyze and replay collective logs obtained from RCCL (ROCm Communication Collectives Library) runs. It can be a useful tool when trying to recreate problem situations (without as much setup), or as a user-directed utility to run collectives (by crafting their own 'logfile').
+
+## Features
+
+- Parses and validates collective logs from RCCL runs.
+- Detects missing/faulty group calls and provides report.
+- Replays collective calls based on the recorded data.
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/README.md:20]: We should point out that we do not record data.  We only record which calls were made and use dummy data.

File: tools/rccl_replayer/README.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+3. **Group Call Validation:** After acquiring data from the collective logs and generating group calls, the replayer validates the results using two different methods. For Non-Send/Recv collectives, it checks if each MPI rank has the required number of collective tasks. For Send/Recv collectives, it verifies if they all have a matching pair.
+
+4. **Replaying RCCL:** Based on the aggregated and validated data, Replayer will replay the collective logs to reproduce the RCCL runs from your application.
+
+5. **Reporting and Skipping.** Replayer outputs the detected faulty group calls and skips them during replay. It provides a report showing which group calls were skipped and why and, at the end, summarizes how many group calls were replayed and how many were skipped.
+
+## Installation
+
+To build the replayer, go to the rccl_replayer directory and simply run 'make'.
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/README.md:41]: not just simply run make - need to provide path to MPI_DIR and then run make.

File: tools/rccl_replayer/README.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+After successfully building the replayer, you can run it using the following command:
+
+```bash
+    mpirun -np <numProcesses> ./rcclReplayer </path/to/logfile> <numGpusPerMpiRank>
+```
+
+Replace <numProcesses> with the number of MPI processes you want to run during the replay, </path/to/logfile> with the path to the collective log file generated during your RCCL runs, and <numGpusPerMpiRank> with the number of GPUs per MPI rank used in your application.
+
+Depending on the MPI library you use, you may need to modify the mpirun command accordingly. The flag NCCL_DEBUG_SUBSYS=COLL ensures that only collective log information is printed to the terminal.
```
Comment (gilbertlee-amd) [inline tools/rccl_replayer/README.md:61]: You didn't remove this COLL line here.

**Review by gilbertlee-amd — CHANGES_REQUESTED** [review]

**Review by gilbertlee-amd — APPROVED** [review]
